Feature: Automatic Stem Exporting

seangoodvibes opened the pull request:
# Automated Stem Exporting

https://github.com/SynthstromAudible/DelugeFirmware/assets/138174805/044efe71-ac49-4c4b-bd5c-92b8eb269582

## Description

Added `STEM EXPORT`, an automated process for exporting `CLIP STEMS` while in `SONG VIEW` and `INSTRUMENT STEMS` while in `ARRANGER VIEW`. Press `SAVE + RECORD` to start exporting stems.

Now with one quick action you can start a stem export job, walk away from your Deluge and come back with a bunch of stems for all your clips and arranger instruments.

## Stems Folder

Stems get exported to a new `SAMPLES/STEMS/` folder. 

Within the `STEMS` folder, a folder with the `SONG NAME` is created for each stem export job which contains all the WAV file recordings. 

If the same SONG is exported more than once, a 5 digit number incremental number is appended to that song's folder name.

## Stem File Names

Stem's are given a meaningful name in the following format:

`ClipType_ExportType_PresetName_FileNumber.WAV`

> For example:
> 
> SYNTH_CLIP_PRESETNAME_00000.WAV
> SYNTH_TRACK_PRESETNAME_00000.WAV

## Shortcuts to Start/Stop Stem Exporting

### Starting Stem Export

- Hold `SAVE` + Press `RECORD` while Playback and Record are disabled to launch Stem Export process
- When the stem export is finished, a dialog will appear on the display that tells you that the stem export process has finished. Press `SELECT`, `BACK` or any `PAD` on the grid to exit the dialog.
- Note 1: In Arranger View, any Instruments that are `MUTED` will be excluded from the stem export.
- Note 2: Stems are exported without Master (Song) FX applied

### Cancelling Stem Export

- Press `BACK` to cancel Stem Export process
- When you cancel stem export, a dialog will appear on the screen asking you to confirm if you want to cancel the export. Press on the `SELECT` encoder to confirm that you want to cancel. Press `BACK` or any `PAD` on the grid to exit out of the dialog and continue with stem export process.
- Note: When you press `BACK`, the stem export still continues in the background until you confirm you want to stop.

Addresses: https://github.com/SynthstromAudible/DelugeFirmware/discussions/2254

## Videos

https://github.com/SynthstromAudible/DelugeFirmware/assets/138174805/77bb8dfc-8b39-408d-8688-fc43eb7be593

https://github.com/SynthstromAudible/DelugeFirmware/assets/138174805/c48d0ab8-656f-427a-99c0-f7c4076b06ca

https://github.com/SynthstromAudible/DelugeFirmware/assets/138174805/c00e69c6-96a6-4b25-94ed-abcf1a0088f4

sichtbeton commented:
First of all: this is amazing. As far as I remember, this feature is the most requested of alle times in the vicinity of teh Deluge, by far. So thank you so much for doing this! Excellent:

Bugs i found so far:
- a sample that started right at beginning was dropped, so it only got recorded after 1 bar
- the mute buttons were not returned to the previous state

a couple of thoughts / improvements:
- a final screen after export that show a little status report and doesnt go away, bc when you leave your deluge on for a complete song with maybe 30 minutes export time in total, you want to be sure, that everything was completed
- export to subfolder
- export with filenames that match the tracks / clips names (and even the song) for easier finding and reference
- export the little export status screen into the folder as a simple text file, so you can quickly see what has been exported without needing to manually check every file

but, besides the little bug not recording from the very first second on, even in this state it already is a MAJOR improvement for exporting tracks.

my prayers have been anwered.

thx!

seangoodvibes commented:
Feedback from @trappar for future PR

I feel like a robust implementation for this should eventually support things like:

* A menu that pops up when batch export is triggered with export options
* Ability to mark/select/deselect tracks for export
* Ability to optionally turn off various things like FX (reverb, delay, etc) and compression during export. Would need to temporarily set all these values and then restore them.
* Disable sound output while exporting.
* Ability to select an output directory and maybe auto name files in some slightly better way like including the preset name and an index.
* Export functionality accessible via menu in addition to button combo.

sichtbeton commented:
And if this all wasnt too much of a wishlist:

- a way to disable normalization while exporting to not mess up volume levels
- mute a track in arranger to be excluded from exporting

seangoodvibes commented:
Another request on Facebook:
- configurable delay between exports to multi track record audio out of the deluge into an external device (eg daw)

trappar commented:
Re: normalization

If you’re able to disable that I’d definitely make that an option rather than always disabling it. Personally if I was going to export stems for a regular DAW I’d prefer the samples be normalized so that I could do my mixing in the daw rather than baking mixing into samples.

I guess if the thing being exported is clips, I'd expect normalization. If it's exporting the arrangement (stems) I would not expect normalization. Perhaps these are really two different features?

sichtbeton commented:
Yep think so too, it should be optional. With for exporting stems and have the mix be kept intact or normalized for maximum levels.

seangoodvibes commented:
> Re: normalization
> 
> If you’re able to disable that I’d definitely make that an option rather than always disabling it. Personally if I was going to export stems for a regular DAW I’d prefer the samples be normalized so that I could do my mixing in the daw rather than baking mixing into samples.
> 
> I guess if the thing being exported is clips, I'd expect normalization. If it's exporting the arrangement (stems) I would not expect normalization. Perhaps these are really two different features?

I think a Normalization menu setting that has 4 options would work:

- OFF - no normalization added at all to stem exports
- CLIP - only adds normalization to stem export for clips
- TRACK - only adds normalization to track export for arranger instruments
- ALL - adds normalization to both clip and track stem exports

m-m-adams reviewed (commented):
With a change to yield instead of using the timer manager I think this looks good!

seangoodvibes commented on src/deluge/playback/playback_handler.cpp:
```diff
@@ -276,6 +278,71 @@ void PlaybackHandler::recordButtonPressed() {
+/// instruments / clips for exporting
+void PlaybackHandler::startStemExportProcess() {
+	// exit save UI mode and turn off save button LED
+	exitUIMode(UI_MODE_HOLDING_SAVE_BUTTON);
```
Will check! Shouldn't be possible but I'll confirm :D

m-m-adams reviewed (commented):
One small nit, looks good!

seangoodvibes commented on src/deluge/storage/audio/audio_file_manager.cpp:
```diff
@@ -256,8 +257,15 @@ Error AudioFileManager::getUnusedAudioRecordingFilePath(String* filePath, String
-
+	// if stem export process has started and this function is called,
+	// then we want to export stem files to a specific stems folder
+	if (stemExport.processStarted) {
```
> can you split all the stem export stuff to it's own function? It's hard to follow the mix of stem and normal export in the control flow

yup will do!

m-m-adams commented on src/deluge/processing/stem_export/stem_export.cpp:
```diff
@@ -427,84 +427,49 @@ void StemExport::displayStemExportProgress7SEG() {
 
-	// wavFileNameForStemExport = "/OutputType_
-	error = wavFileNameForStemExport.concatenate("_");
+	error = stemExport.getUnusedStemRecordingFolderPath(filePath, folder);
```
should just be getUnusedStemRecordingFolderPath and not refer to a global object

m-m-adams commented on src/deluge/processing/stem_export/stem_export.cpp:
```diff
@@ -427,84 +427,49 @@ void StemExport::displayStemExportProgress7SEG() {
-		error = wavFileNameForStemExport.concatenate("CLIP_");
+	// wavFileName is uniquely set for each stem export
+	// when this flag is true, there is a valid wavFileName that has been set for stem exporting
+	if (stemExport.wavFileNameForStemExportSet) {
```
same for all of this - they shouldn't be prefixed with stemExport.

alifeinbinary commented:
Monumental PR. Thank you @seangoodvibes !

seangoodvibes commented:
> Hi, first of @seangoodvibes all thanks for implementing stem-export! I have a question/ problem though: I understood that when you are in arranger mode, and start the process, the whole arrangement of each instument would get exported in individual wav-files, but when I try it, I only get one bar of each instrument instead of im my case 5:20 min long wave files, which means I would have to recreate the arrangement in an DAW to have the complete arranged track. So what did I get this wrong, or is there another way to manage to get the full arrangement with the individual wave-files? thanks in advance

Hey, that shouldn't be happening. As you expected, it should be exporting each instrument up to the length of the arrangement. Other people have reported a similar bug. Are you able to share a collect all saved version of your song so I can try and see what's going on?

Can you confirm if this also happens with a new song?

Thanks

feilsch-hering commented:
Sorry I'm not sure what you mean with "collect all saved version". If you explain a bit more what exactly you need, I can try.
Anyways I started the track with a beta of community 1.1 if I remember right (5th of june) or maybe it was still Community 1.0, not sure andy I updated to 1.1 while working on it.  I tried the stem-export with the todays nightly and previously a nightly from ~2 weeks ago, both times I only had one bar of the instruments, which was strange because the rendering took quite long. Rendering of only one bar for each instrument shouldf have been much faster, so it felt the whole song was rendered but the files written are just one bar.
I haven't try yet to create a new song from the nightlys yet and I would of course also hope for a working export of songs started in previous versions. But if it's important I can make a quick new one and report back.

seangoodvibes commented:
> Sorry I'm not sure what you mean with "collect all saved version". If you explain a bit more what exactly you need, I can try. Anyways I started the track with a beta of community 1.1 if I remember right (5th of june) or maybe it was still Community 1.0, not sure and I updated to 1.1 while working on it. I tried the stem-export with the todays nightly and previously a nightly from ~2 weeks ago, both times I only had one bar of the instruments, which was strange because the rendering took quite long. Rendering of only one bar for each instrument should have been much faster, so it felt the whole song was rendered but the files written are just one bar. I haven't try yet to create a new song from the nightlys yet and I would of course also hope for a working export of songs started in previous versions. But if it's important I can make a quick new one and report back.

Hey, so your description gives me a clue.

You said the recording took very long. Did it continue recording after the playhead reached the end of the instrument track? If so, it's possible you're experiencing a problem similar to what someone else reported.

Basically, by default, the stem export will record the full length of the arrangement, but it will also keep recording afterwards if the sound has not yet dropped to zero. This allows it to record tails in the stems. 

There seems to be a scenario, which may or may not be a bug, where the tails keep playing and the sound does not fully drop to zero.

if you turn on the VU meter before starting the stem export, you can see whether or not the VU meter drops to zero at the end of the recording.

You can also try the release build from here:

https://github.com/SynthstromAudible/DelugeFirmware/actions/runs/10435480991?pr=2456

I've added an option in the SONG > STEM EXPORT > CONFIGURE menu where you can turn off EXPORT TO SILENCE

By doing that, it will not record tails and it will finish recording the stem at the end of the arrangement length.

Could you try and see if turning off "Export to Silence" solves your issue?

Thanks

feilsch-hering commented:
I tried with a new song now, and with a 4 bar arrangement I have now a 16 seconds long wave with a filesize of ~2mb. Previously I had filesize of 88MB but only 5 seconds even though it should have been 5:20 min, which is super odd.
I cant tell you what happened with the playhead at the end but I think it correctly changed to the next instument at the end. The track has not very long tails, maybe a few seconds the most from the delays.
I have to bring my kid to bed now which could take ~2 hours, afterwards I can try again with th eVUmeter engaged. And also by expoirting from the menu.

feilsch-hering commented:
I now tried with the VU meter and cross-screen engaged in order to see the behaviour at the end, but it pretty much directly at the end of the song jumps to the next instrument and at that moment VU shows silence. The reason that exporting takes long in my case, is that the song has 17 instruments and each one takes more than i minute to render, so it's about 20 min for one run.
In the current nightly I only see the Normalisation option. I haven't looked into making a custom-build yet, so I can't try the new option now. Please let me know if it's included in the nightly, then I can try again.
But like I wrote, I always get 82.7MB Wave files, which only contain a 5 seconds sound, so something is wrong with the way the waves are written.

feilsch-hering commented:
Just to let you know, tonight I tried again with deluge-v1_2_0-beta+2024_09_02-6d9319b and afterwards with deluge-v1_2_0-nightly+2024_09_02-8d0d66d and my problem is still the same, all 80+ mb wavs play only 5 seconds of audio.

seangoodvibes commented:
> Just to let you know, tonight I tried again with deluge-v1_2_0-beta+2024_09_02-6d9319b and afterwards with deluge-v1_2_0-nightly+2024_09_02-8d0d66d and my problem is still the same, all 80+ mb wavs play only 5 seconds of audio.

Did you post in our discord earlier today about this?

if not, I'll suggest the same thing I suggested to that person:

1) can you test with record to silence turned off in the stem export config menu. Does it make a difference?

2) are you using the "freeverb" reverb model in your song or the "mutable" reverb model? There is currently a bug with the "freeverb" reverb model that hasn't been fixed yet. If you are using "freeverb" can you try changing to "mutable" and see if that fixes your problem?

If changing to "mutable" fixes your problem, can you try using "mutable" with "record to silence" turned on as well to confirm you can export successfully with record to silence as well.

Thanks

feilsch-hering commented:
No I didn't post on Discord but on the Facebook-group about it. If you reread our above back and forth posts about the issue, I already tried your suggestions without that I managed to get a different outcome. And I'm using Mutable Reverb, so that shouldn't cause this issue. I tried engaging Stem-render per menu, turned off to silence, looked at the VU meter and it goes to zero at the end of the track, tried with and without FX, I always end up with ~80mb big wave-files with only 5 seconds playing.
In case you want to take a look at one of my 80mb/5sec wave files I could upload one for you for example with wetransfer.

seangoodvibes commented:
> No I didn't post on Discord but on the Facebook-group about it. If you reread our above back and forth posts about the issue, I already tried your suggestions without that I managed to get a different outcome. And I'm using Mutable Reverb, so that shouldn't cause this issue. I tried engaging Stem-render per menu, turned off to silence, looked at the VU meter and it goes to zero at the end of the track, tried with and without FX, I always end up with ~80mb big wave-files with only 5 seconds playing. In case you want to take a look at one of my 80mb/5sec wave files I could upload one for you for example with wetransfer.

It sounds like you have the same issue as reported here https://github.com/SynthstromAudible/DelugeFirmware/issues/2573

Can you please upload a collect all version of your song to that issue (you can drag and drop a file into the comments box).

Can you also please confirm if this issue happens with all your songs or just this one.

Thank you

feilsch-hering commented:
Yes, #2573 is exactly the bug I discussed with you 2 weeks ago. 
How can I do a "collect all"? I already asked for a hint 2 weeks ago, otherwise I would have uploaded it for you already.
I tried with 2 different songs and both have the 5 seconds issue.
OK, shall we continue to discuss the bug here, or continue on the #2573 threat?

seangoodvibes commented:
> Yes, #2573 is exactly the bug I discussed with you 2 weeks ago. How can I do a "collect all"? I already asked for a hint 2 weeks ago, otherwise I would have uploaded it for you already. I tried with 2 different songs and both have the 5 seconds issue. OK, shall we continue to discuss the bug here, or continue on the #2573 threat?

lets continue on #2573 please

you can collect all and save as follows:

> press save, then press and hold select, it'll save the whole project into a folder 

Thanks
